[image_picker_macos] macOS native image picker

EchoEllet opened the pull request:
*Adds support for macOS native image picker for `image_picker_macos`.*

An alternative implementation that is based on [PHPickerViewController](https://developer.apple.com/documentation/photokit/phpickerviewcontroller) instead of [file_selector](https://pub.dev/packages/file_selector).

https://github.com/user-attachments/assets/b4cfd853-648a-4a1f-90d4-d0dc6322fbbb

<img src="https://github.com/user-attachments/assets/78ec9a1e-46e0-4758-a9d7-05718bdb8f8f" alt="macOS native image picker window" width="600" />

## Usage

Similar to the [Android Photo Picker in `image_picker_android`](https://pub.dev/packages/image_picker_android#photo-picker):

```dart
import 'package:image_picker_macos/image_picker_macos.dart';
import 'package:image_picker_platform_interface/image_picker_platform_interface.dart';
// ···
  final ImagePickerPlatform imagePickerImplementation =
      ImagePickerPlatform.instance;
  if (imagePickerImplementation is ImagePickerMacOS) {
    imagePickerImplementation.useMacOSPHPicker = true;
  }
```

> [!NOTE]
> It's only supported on macOS +13, on macOS 12 and older versions, will fallback to the [`file_selector_macos`](https://pub.dev/packages/file_selector_macos) implementation.

## Related Issues

- *Fix https://github.com/flutter/flutter/issues/125829*
- *Fix https://github.com/flutter/flutter/issues/150063*
- *Related https://github.com/flutter/flutter/issues/70234*
- *Related https://github.com/flutter/packages/pull/4172*
- *Related https://github.com/flutter/packages/pull/3882*
- *Related https://github.com/flutter/flutter/issues/85100*

## Limiations

* Supports `maxWidth`, `maxHeight`, `imageQuality`, and `limit` when using the PHPicker implementation.
* Supports picking image, multi-image, video, media, multi-media. The multi-media supports picking images and videos.
* Doesn't support `maxDuration` of video selection, and `requestFullMetadata` (iOS-specific).
* The user can only pick photos from the macOS Photos app. The user will see a [`No Photos` message](https://discord.com/channels/608014603317936148/1295165633931120642/1295851244132499506) if they don't have photos within the app.

## Motivation

The macOS native picker only supports picking images, videos, and media from the [Photos for macOS](https://www.apple.com/in/macos/photos/) app.

If the user has photos in a directory, they will need to open the app and import the images they need, however, this app integrates with other macOS apps and the Apple ecosystem, and it supports features such as importing pictures seamlessly from an iOS device, albums, favorites, recently deleted, duplicates.

<details><summary>Screenshot</summary>
<p>

![Importing iOS photos to the macOS Photo Picker](https://github.com/user-attachments/assets/9156e8f4-84c0-46e7-8744-9ea88b712979)

</p>
</details> 

> [!TIP]
> It doesn't require [file read-only access entitlement](https://pub.dev/packages/file_selector_macos#entitlements).

### IMO

it's better to keep this implementation optional instead of making it as default, the developer can provide the option to switch between implementations based on the user preference since it's not common on desktops to have a photos app like mobile platforms, maybe provide a method to check if it's supported:

```dart
ImagePicker().supportsPHPicker();

// OR

final ImagePickerPlatform imagePickerImplementation =
      ImagePickerPlatform.instance;
  if (imagePickerImplementation is ImagePickerMacOS) {
    imagePickerImplementation.supportsPHPicker();
  }
```

## Testing

This branch has been manually tested on **macOS 14.6.1**.
Native and Dart Unit tests have been added, though unfortunately, native UI tests are missing (see [#70234](https://github.com/flutter/flutter/issues/70234)). I left a TODO since currently it doesn't seem to be possible.

## Additional Context

**Related Discord Threads**:

- *[Unable to create native macOS Unit/UI tests due to incompatibility between Xcode 16 and CocoaPods](https://discord.com/channels/608014603317936148/608020293944082452/1295165633931120642)*
- [*Unable to tap a Flutter widget button from macOS `XCUITest`*](https://discord.com/channels/608014603317936148/608020293944082452/1300517990957056080)

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I [linked to at least one issue that this PR fixes] in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style], or this PR is [exempt from CHANGELOG changes].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[Discord]: https://github.com/flutter/flutter/blob/master/docs/contributing/Chat.md
[linked to at least one issue that this PR fixes]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md#overview
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#version
[following repository CHANGELOG style]: https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#changelog-style
[exempt from CHANGELOG changes]: https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#changelog
[test-exempt]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md#tests

EchoEllet commented on packages/image_picker/image_picker_macos/example/macos/RunnerTests/ImageCompressTests.swift:
```diff
@@ -0,0 +1,37 @@
+
+final class ImageCompressTests: XCTestCase {
+
+  private func createTestImage(size: NSSize) -> NSImage {
+    let image = NSImage(size: size)
+    image.lockFocus()
+    NSColor.white.set()
+    NSBezierPath(rect: NSRect(origin: .zero, size: size)).fill()
+    image.unlockFocus()
+    return image
+  }
```
This is duplicated in both `ImageCompressTests` and `ImageResizeTests`, should I extract it?

stuartmorgan replied:
Yes, duplicate test support code should be extracted.

> Most unit tests of plugins on this repo are implemented with one file `RunnerTests`.

Many plugins are either small enough that we didn't need multiple files, or are significantly under-tested on the native side (native unit test integration having been added after the fact in many plugin); the single file approach isn't a specific outcome we are going for.

EchoEllet commented on packages/image_picker/image_picker_macos/example/macos/Runner/AppDelegate.swift:
```diff
@@ -5,9 +5,13 @@
   override func applicationShouldTerminateAfterLastWindowClosed(_ sender: NSApplication) -> Bool {
     return true
   }
+
+  override func applicationSupportsSecureRestorableState(_ app: NSApplication) -> Bool {
+    return true
+  }
```
This has been automated somehow, should I revert?

EchoEllet commented on packages/image_picker/image_picker_macos/example/macos/Runner/AppDelegate.swift:
```diff
@@ -5,9 +5,13 @@
   override func applicationShouldTerminateAfterLastWindowClosed(_ sender: NSApplication) -> Bool {
     return true
   }
+
+  override func applicationSupportsSecureRestorableState(_ app: NSApplication) -> Bool {
+    return true
+  }
```
This has been automated somehow, should I revert?

EchoEllet replied:
This is due to running `flutter run -d macos` when using the `main` channel (automated):

```console
macos/Runner/AppDelegate.swift does not override
applicationSupportsSecureRestorableState. Updating.
```

EchoEllet commented on packages/image_picker/image_picker_macos/example/macos/RunnerTests/ImageResizeTests.swift:
```diff
@@ -0,0 +1,236 @@
+
+final class ImageResizeTests: XCTestCase {
+
+  private func createTestImage(size: NSSize) -> NSImage {
+    let image = NSImage(size: size)
+    image.lockFocus()
+    NSColor.black.set()
+    NSBezierPath(rect: NSRect(origin: .zero, size: size)).fill()
+    image.unlockFocus()
+    return image
+  }
```
Duplicated.

EchoEllet commented on packages/image_picker/image_picker_macos/lib/image_picker_macos.dart:
```diff
@@ -24,6 +30,42 @@ class ImagePickerMacOS extends CameraDelegatingImagePickerPlatform {
+  /// The [PHPicker](https://developer.apple.com/documentation/photokit/phpickerviewcontroller)
+  /// is **supported on macOS 13.0+**
+  @visibleForTesting
+  Future<bool> supportsPHPicker() => _hostApi.supportsPHPicker();
```
This was initially private before modifying the unit tests, though should it be public in both cases?

EchoEllet commented on packages/image_picker/image_picker_macos/lib/image_picker_macos.dart:
```diff
@@ -24,6 +30,42 @@ class ImagePickerMacOS extends CameraDelegatingImagePickerPlatform {
+  ///  * [useMacOSPHPicker] to check whether **PHPicker** should be preferred over [file_selector_macos](https://pub.dev/packages/file_selector_macos).
+  ///  * [supportsPHPicker] to verify if the current macOS version supports **PHPicker**.
+  @visibleForTesting
+  Future<bool> shouldUsePHPicker() async =>
```
https://github.com/flutter/flutter/blob/master/docs/contributing/Style-guide-for-Flutter-repo.md#avoid-using-visiblefortesting

Does mean that this should be public?

stuartmorgan replied:
We don't follow that in flutter/packages; I should update it to node that. We do use `@visibleForTesting`, and we do not consider it public API for semver purposes (a distinction which is irrelevant in flutter/flutter).

EchoEllet commented on packages/image_picker/image_picker_macos/lib/image_picker_macos.dart:
```diff
@@ -83,8 +125,9 @@ class ImagePickerMacOS extends CameraDelegatingImagePickerPlatform {
 
-  // [ImagePickerOptions] options are not currently supported. If any
-  // of its fields are set, they will be silently ignored.
+  // [ImagePickerOptions] options are currently only supported when using
+  // PHPicker implementation. If any of its fields are set,
+  // they will be silently ignored.
```
Nit: This message and the others could be improved for clarity.

EchoEllet commented on packages/image_picker/image_picker_macos/macos/image_picker_macos/Sources/image_picker_macos/ImageCompress.swift:
```diff
@@ -0,0 +1,71 @@
+    guard let tiffData = self.tiffRepresentation,
+      let bitmapRep = NSBitmapImageRep(data: tiffData)
+    else {
+      // TODO(EchoEllet): Is there a convention for the error code? ImageConversionError or IMAGE_CONVERSION_ERROR or image-conversion-error. Update all codes.
+      throw PigeonError(
+        code: "ImageConversionError", message: "Failed to convert NSImage to TIFF data.",
```
Probably `image-conversion-error`?

EchoEllet replied:
Is there a better way to handle error codes? Maybe an enum in the `pigeons/messages.dart` file. Currently, they are not being handled from the Dart side as it's pretty rare to encounter any of them (e.g., not enough storage or an unknown error while writing the temp file).

tarrinneal replied:
@stuartmorgan for conventions.

Any error thrown should be handled on the dart side as per policy.

stuartmorgan replied:
https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#platform-exception-handling

Known string constants that for the error code (matching on both sides of the communication channel) are an option, although of course that's the kind of thing that we switched to Pigeon to avoid so it's not ideal. In some packages where we've started moving to the new guidance, we do indeed use a message.dart enum so that we can do exhaustive, type-safe handling on the Dart side. In that model we use a return class that can include error information, and don't use `PigeonError` at all.

EchoEllet replied:
> In that model we use a return class that can include error information, and don't use `PigeonError` at all. (`url_launcher_macos` is [a simple example of that approach](https://github.com/flutter/packages/blob/main/packages/url_launcher/url_launcher_macos/pigeons/messages.dart).)

In case the API call is `@async`, the `completion` expects a Swift `Result` that's either a `failure` or `success`. The `failure` requires a class that conforms to `Error`, so the return class generated by Pigeon can't be used. Is it okay to use `success` instead to return the error result?

Throwing the error from Swift:

```swift
completion(.success(ImagePickerErrorResult(error: .phpickerUnsupported)))
```

Handling the error from Dart:

```dart
final ImagePickerResult result = await _hostApi.pickImages(
    ImageSelectionOptions(quality: 100), GeneralOptions(limit: 1));
switch (result) {
  case ImagePickerSuccessResult():
    final List<String> filePaths = result.filePaths;
  case ImagePickerErrorResult():
    switch (result.error) {
      case ImagePickerError.phpickerUnsupported:
        throw UnsupportedError(
            result.platformErrorMessage.toString());
      // Exhaustive type-safe handling...
    }
}
```

I'm still uncertain what to throw after handling those platform errors in a type-safe way from the Dart side. Should I throw built-in exceptions from Dart whenever possible (e.g., `ArgumentError`, `UnsupportedError`)? As I didn't find any appropriate error types defined by the platform interface package.

stuartmorgan replied:
> Is it okay to use `success` instead to return the error result?

That is currently the only way to do error handling without string matching, yes. (TBD if we will eventually build some alternate error system into Pigeon for this pattern.)

> I'm still uncertain what to throw after handling those platform errors in a type-safe way from the Dart side. Should I throw built-in exceptions from Dart whenever possible (e.g., `ArgumentError`, `UnsupportedError`)?

`ArgumentError` and `UnsupportedError` are errors, not exceptions. It would be incorrect in Dart to throw an error unless the package client did something wrong that should have been fixed at the time of authoring the client code. Runtime failures that are not the developer's fault (e.g., bad image data on a user's device) must be exceptions, not errors.

> As I didn't find any appropriate error types defined by the platform interface package.

For now, it's fine to throw a `PlatformException` on the Dart side, with a TODO comment saying that this is for compatibility with the existing behavior of other implementations of `image_picker`, and that it should be replaced with a plugin-specific exception as part of a cross-platform update.

EchoEllet replied:
> it's fine to throw a PlatformException on the Dart side
> In that model we use a return class that can include error information, and don't use PigeonError at all. 

Done in 2a101a9ab613a9fbba5a557d8a3eab7f425120d6.

> Known string constants that for the error code (matching on both sides of the communication channel) are an option, although of course that's the kind of thing that we switched to Pigeon to avoid so it's not ideal. 

The `PigeonError` is no longer used and the error handling is type-safe now.

EchoEllet commented on packages/image_picker/image_picker_macos/macos/image_picker_macos.podspec:
```diff
@@ -0,0 +1,23 @@
+                       DESC
+  s.homepage         = 'https://github.com/flutter/packages'
+  s.license          = { :type => 'BSD', :file => '../LICENSE' }
+  s.author           = { 'Flutter Team' => 'flutter-dev@googlegroups.com' }
```
Nit: Some packages use a different name (e.g. [podspec of `image_picker_ios`](https://github.com/flutter/packages/blob/6c4482a64492e1cc5980f856bec36fb4d96fa983/packages/image_picker/image_picker_ios/ios/image_picker_ios.podspec#L14)).

EchoEllet commented on packages/image_picker/image_picker_macos/macos/image_picker_macos/Sources/image_picker_macos/ImageCompress.swift:
```diff
@@ -0,0 +1,71 @@
+  /// - Parameter quality: The quality of the image (0 to 100).
+  /// - Returns: An optional `NSImage` that represents the compressed image.
+  func compressed(quality: Int64) throws -> NSImage {
+    assert(quality != 100, "Quality 100 means no compression.")
+    assert(quality >= 0, "Quality can't be negative.")
```
Should I remove the asserts from the platform side since it will crash the app when running using `flutter run -d macos`?

stuartmorgan replied:
Argument assertions should be happening via `ArgumentError` checks on the Dart side.

EchoEllet replied:
> assert(quality != 100, "Quality 100 means no compression.")

This check is unnecessary since `shouldCompressImage` will return `false` if the quality is `100`.

> assert(quality >= 0, "Quality can't be negative.")

```diff
ImageSelectionOptions _imageOptionsToImageSelectionOptions(
    ImageOptions imageOptions,
  ) {
+    final int? imageQuality = imageOptions.imageQuality;
+    if (imageQuality != null && imageQuality < 0) {
+      throw ArgumentError.value(
+          imageQuality, 'imageQuality', 'quality cannot be negative.');
+    }
    return ImageSelectionOptions(
      quality: imageQuality ?? 100,
      maxSize: MaxSize(
        width: imageOptions.maxWidth,
        height: imageOptions.maxHeight,
      ),
    );
  }
```

EchoEllet commented on packages/image_picker/image_picker_macos/README.md:
```diff
@@ -25,14 +55,18 @@ should add it to your `pubspec.yaml` as usual.
 
 ### Entitlements
 
-This package is currently implemented using [`file_selector`][3], so you will
-need to add a read-only file acces [entitlement][4]:
+This package’s default implementation relies on [file_selector][3],
+which requires the following read-only file access entitlement:
```
Nit: I will restore the backticks.

EchoEllet commented on packages/image_picker/image_picker_macos/example/macos/Runner.xcodeproj/xcshareddata/xcschemes/Runner.xcscheme:
```diff
@@ -1,10 +1,28 @@
    <BuildAction
       parallelizeBuildables = "YES"
       buildImplicitDependencies = "YES">
+      <PreActions>
+         <ExecutionAction
+            ActionType = "Xcode.IDEStandardExecutionActionsCore.ExecutionActionType.ShellScriptAction">
+            <ActionContent
```
Xcode project changes are mostly automated except https://github.com/flutter/flutter/blob/master/docs/ecosystem/testing/Plugin-Tests.md#enabling-xctests-or-xcuitests

EchoEllet reviewed (commented):
I left TODOs in the code that should be addressed in this PR before continuing further except the [TODO of adding macOS native UI tests](https://github.com/EchoEllet/packages/blob/feat/phpicker-image-picker-macos/packages/image_picker/image_picker_macos/example/macos/RunnerUITests/RunnerUITests.swift#L29-L30).

I welcome any feedback and suggestions.

EchoEllet commented:
> Mac_arm64 macos_repo_checks
> Tasks failed: Swift format,validate iOS and macOS podspecs

All Swift files are formatted except the generated `Messages.g.swift`. I ran [`dart run script/tool/bin/flutter_plugin_tools.dart format --packages image_picker_macos
`](https://github.com/flutter/packages/blob/main/script/tool/README.md#format-code).

EchoEllet commented:
The docs of `image_picker` have not been updated.

> [temporary file could not be created (iOS only)](https://github.com/flutter/packages/blob/main/packages/image_picker/image_picker/lib/image_picker.dart#L70)
> Future<XFile?> pickImage

This is also possible with the native macOS implementation (when PHPicker is enabled). Should I update those parts of the docs to reflect the change?

EchoEllet commented:
The `ImageSource` is unsupported for both implementations, should override the `supportsImageSource`?

```dart
@override
bool supportsImageSource(ImageSource source) {
  if (source == ImageSource.camera) {
    return cameraDelegate != null;
  }
  return false;
}
```

EchoEllet commented on packages/image_picker/image_picker_macos/macos/image_picker_macos/Sources/image_picker_macos/ImagePickerImpl.swift:
```diff
@@ -0,0 +1,519 @@
+      noActiveWindow()
+      return
+    }
+    // TODO(EchoEllet): IMPORTANT The window size of the picker is smaller than expected, see the video in https://discord.com/channels/608014603317936148/1295165633931120642/1295470850283147335
```
This is an important issue to fix since it affects the user experience, see the video in this PR.

EchoEllet commented on packages/image_picker/image_picker_macos/lib/image_picker_macos.dart:
```diff
@@ -141,30 +202,77 @@ class ImagePickerMacOS extends CameraDelegatingImagePickerPlatform {
+              _imageOptionsToImageSelectionOptions(options.imageOptions),
+        ),
+        GeneralOptions(
+          limit: options.limit ?? (options.allowMultiple ? 0 : 1),
```
This seems to be incorrect, and needs validation [similarly to image_picker_android](https://github.com/flutter/packages/blob/main/packages/image_picker/image_picker_android/lib/image_picker_android.dart#L275-L318) since the current code doesn't check `allowMultiple` in case `limit` was not provided.

EchoEllet replied:
It looks like [there is validation from the app-facing package](https://github.com/flutter/packages/blob/d8e8e8cee712b4bea345b804c9fc88df229ce154/packages/image_picker/image_picker/lib/image_picker.dart#L246-L265) (`image_picker`) for all methods, do we still need to verify from the platform implementation side?

stuartmorgan replied:
Ideally [this](https://github.com/flutter/packages/blob/a13b4ef628f634bbbd4a5dcfb96f16a6500578a7/packages/image_picker/image_picker_platform_interface/lib/src/types/media_options.dart#L34-L35) would document that `allowMultiple` can be ignored if `limit` is set, but since we expect all clients in practice to be using `createAndValidate` it's fine to assume that even though it's not currently documented.

cbracken commented:
> @cbracken Is the triage-macos label being reviewed in triage?

Yes, but we're unfortunately quite backlogged and we prioritise the iOS review. Given how far back this is, it's probably worth me scheduling an extra catch-up triage.

I realise this isn't a viable long-term solution, but in the meantime, if there are any specific patches you need to get unblocked, feel free to ping me directly on them.

jmagman commented:
Adding @cbracken as a reviewer.  Sorry again for the delay on this one.

stuartmorgan commented on packages/image_picker/image_picker_macos/lib/image_picker_macos.dart:
```diff
@@ -24,6 +30,42 @@ class ImagePickerMacOS extends CameraDelegatingImagePickerPlatform {
+  ///
+  /// The [PHPicker](https://developer.apple.com/documentation/photokit/phpickerviewcontroller)
+  /// is **supported on macOS 13.0+**
+  @visibleForTesting
```
The PR description suggests that developers would want to provide this as a user-level opt-in, in which case they would need this information to know whether to show the option, so this should be public.

stuartmorgan commented:
> I ran the script [`dart run script/tool/bin/flutter_plugin_tools.dart format --packages image_picker_macos `](https://github.com/flutter/packages/blob/main/script/tool/README.md#format-code) to format the `Messages.g.swift` though the output produced some warnings. Should they ignored?

Yes, that's just a Pigeon output style issue.

> Formatted using `dart format` since the [format tool within this repo](https://github.com/flutter/packages/blob/main/script/tool/README.md#format-code) didn't format them.

Please file an issue with repro steps; we regularly format Pigeon-generated code with the repo tooling.

> I'm using Xcode 16.1 with Swift 6, maybe the CI uses Swift 5?

CI uses Xcode 15.

> The doc comments of [`image_picker`](https://pub.dev/packages/image_picker) have not been updated.
> 
> > [temporary file could not be created (iOS only)](https://github.com/flutter/packages/blob/main/packages/image_picker/image_picker/lib/image_picker.dart#L70)
> > Future<XFile?> pickImage
> 
> This is also possible with the native macOS implementation (when PHPicker is used). Should I update those parts of the docs to reflect the change?

Yes, if this makes old comments in the app-facing package that still use platform-specific references incorrect, the comments should be updated to follow our newer standard of not making prescriptive comments about platforms in app-facing-package code comments.

> The `ImageSource` is unsupported for both implementations, should override the `supportsImageSource`?

I don't understand the question; does this PR break the camera delegation system for some reason? If so, that would need to be fixed before this can land; we aren't going to regress that behavior to get the new picker (nor should we need to).

> > The arguments `maxWidth`, `maxHeight`, `imageQuality`, and `limit` are only supported when using the [PHPicker](#phpicker) implementation;
> 
> It's also possible to support those on `file_selector` implementation using the same platform API, should we support resizing and compressing for `file_selector` implementation?

I'm not following this one either. What `file_selector` resizing and compression options are you referring to?

EchoEllet commented:
> I'm not following this one either. What `file_selector` resizing and compression options are you referring to?

The arguments `maxWidth`, `maxHeight`, `imageQuality`, and `limit` are only supported when using the new picker. It's possible to support those arguments even when `useMacOSPHPicker` is `false`. The image file will be on disk after being picked using `file_selector` (related #150063). Should we support or keep those arguments only supported when using this new implementation?

> does this PR break the camera delegation system for some reason?

No, I was double-checking in case something was outdated.

> f this makes old comments in the app-facing package that still use platform-specific references incorrect

They are slightly outdated but not incorrect. The [temporary file could not be created error](https://github.com/flutter/packages/blob/main/packages/image_picker/image_picker/lib/image_picker.dart#L70) is only applicable when using the new picker on macOS. Should the new message be `temporary file could not be created (iOS and macOS only)`?

stuartmorgan commented:
> They are slightly outdated but not incorrect. The [temporary file could not be created error](https://github.com/flutter/packages/blob/main/packages/image_picker/image_picker/lib/image_picker.dart#L70) is only applicable when using the new picker on macOS. Should the new message be `temporary file could not be created (iOS and macOS only)`?

The comment should be changed not to be prescriptive about what platforms could throw any given error. It's fine for it to provide examples of errors that can be thrown, but should not claim specific knowledge about specific implementations (all of which are now in different packages from the comment).

stuartmorgan commented:
> > I'm not following this one either. What `file_selector` resizing and compression options are you referring to?
> 
> The arguments `maxWidth`, `maxHeight`, `imageQuality`, and `limit` are only supported when using the new picker. It's possible to support those arguments even when `useMacOSPHPicker` is `false`. The image file will be on disk after being picked using `file_selector` (related [#150063](https://github.com/flutter/flutter/issues/150063)). Should we support or keep those arguments only supported when using this new implementation?

Adding support for image modification in the `file_selector`-based codepath should not be part of this PR if that's what you are asking.

cbracken commented:
Given the current state of Flutter desktop prioritisation, our policy has been to continue to keep desktop stable add quickly address P0 issues, but not to land new feature work or expand scope.

I want to be clear that this is the sort of patch we'd have been absolutely thrilled to have when desktop was a much higher priority, and personally this is a pretty exciting patch, but at present our policy is to avoid expanding scope or landing new feature work for desktop.

@stuartmorgan from the ecosystem perspective, is this something we'd take on if it were community-reviewed and maintained?

stuartmorgan commented:
Apologies @EchoEllet; I didn't realize this was no longer considered in scope or I definitely would have said something much earlier in the process!

If this implementation were published as a new unendorsed `image_picker` implementation package, we could certainly reference it as a potential alternative in the issue about native support and in the [limitations section of the README](https://pub.dev/packages/image_picker_macos#limitations).

EchoEllet commented:
It was great working with you, even if it was just for a short time. Thanks!

> If this implementation were published as a new unendorsed image_picker implementation package
> is this something that might be better maintained as a community package

I’m more than willing to create a separate package—would you be open to reviewing the PR first? It might help decide the best path forward.

> keep desktop stable add quickly address P0 issues, but not to land new feature work or expand scope.
> avoid expanding scope or landing new feature work for desktop.

It seems we're aligned in our opinion.

stuartmorgan commented:
> I’m more than willing to create a separate package—would you be open to reviewing the PR first? It might help decide the best path forward.

Can you elaborate on what you mean by "decide the best path forward"? Code doesn't need to be reviewed by the Flutter team to be published on pub.dev.

EchoEllet commented:
I see, I'm working on the package and will be published soon. Should we close this PR?

EchoEllet commented:
I'm working on the package and will be published soon. Should we close this PR?

EchoEllet commented:
I have extracted this functionality into [native_image_picker_macos](https://github.com/CompileKernel/native-image-picker-macos) as an unendorsed platform implementation, which has not yet been published to pub.dev.

> Can you elaborate on what you mean by "decide the best path forward"?

I'm unfamiliar with native macOS and Flutter macOS integration and there is only [one issue that's critical to fix](https://github.com/CompileKernel/native-image-picker-macos/issues/2) before releasing the package.

I have added [a comment and a TODO previously](https://github.com/flutter/packages/pull/8079#discussion_r1842094973).

> published on pub.dev

This issue should be fixed before publishing the package as stable.

Should we close this PR?

stuartmorgan commented:
> Should we close this PR?

Yes, sorry for the delay on this. I was waiting on an offline discussion to make sure I understand the current scoping decisions.

I'll follow up on the issue.

EchoEllet commented:
No worries. Thank you all for your help with this PR!

EchoEllet commented:
> is this something that might be better maintained as a community package we could point people to in place of our existing implementation?

> If this implementation were published as a new unendorsed image_picker implementation package, we could certainly reference it as a potential alternative in the issue about native support and in the [limitations section of the README](https://pub.dev/packages/image_picker_macos#limitations).

The package [`native_image_picker_macos`](https://pub.dev/packages/native_image_picker_macos) has been published.
